### PR TITLE
feat(ListBox): fix labels of trigger buttons

### LINKED ISSUE #4531: AVT 1 - Dropdown had DAP violations
## Environment
macOS Mojave version 10.14.6
Chrome Version Version 77.0.3865.90
Carbon v10 - React
DAP IBM Accessibility 1.8.0.0 Sept 2019 Ruleset

## Detailed Description
1. Go to the React Component 
2. Run DAP and the following violation is present (see screenshot):
Screen Shot 2019-11-01 at 2 50 43 PM
3. Select the disable knob and run DAP again and the following violation is persent:
Screen Shot 2019-11-01 at 2 51 54 PM
4. Deselect the disable knob and select the form invalid knob and the following violation is present:
Screen Shot 2019-11-01 at 2 54 23 PM

### PR DESCRIPTION
**Pull request by asudoh**
This change fixes an DAP error in `<ListBox.Field>` that happens because its `aria-label` not matching its text content. DAP takes `aria-labelledby` as a preferred label content (over text content or `aria-label`), and thus this change add `aria-labelledby` to `<ListBox.Field>`.

The original `aria-label` content, which is for letting user know that the UI can open/close menu, is replaced with a browser's native approach to tell assistive technology that the button is a pop-up button and let screen reader announce as such.

This change also changes `<label>` to `<span>` for dropdown and non-filterable multi select given there is no form control (`<input>`) associated with it.

Fixes #4531.

#### Changelog

**New**

- A mechanism to map `<ListBox.Field>` to list box label.

**Changed**

- Replaced `<label>` with `<span>` for the labels of dropdown and (non-filterable) multi select.

**Removed**

- `aria-label` from the `<div role="button">` in `<ListBox.Field>`.

#### Testing / Reviewing

Testing should make sure combo box, drop down, multi select, filterable multi select are not broken.

For reviewers - It's easier to review with whitespace change ignored, it can be done by clicking on the gear icon in Files changed tab and check off "Hide whitespace changes".

### CODE REVIEW

**Comment by joshblack** [comment]
One thing, do we even need the `aria-label` to call out `Open menu`? Screen reader software would be able to say whether the listbox is open/closed through `aria-haspopup` and `aria-expanded` if I understand correctly.

**Comment by asudoh** [comment]
This PR removes `aria-label` to call out `Open menu`.

**Comment by joshblack** [comment]
For sure! Sorry if I wasn't clear. I was just thinking through the DAP violation and it seemed like only removing the `aria-label` would be required as the violation seemed to be due to the mismatch of the `aria-label` being `Open menu` but the content being different.

What is the expected behavior for VO if the button/field is associated with the label? Specifically, how are we thinking the value would be announced?

**Comment by asudoh** [comment]
If we just remove `aria-label`, the text content inside the trigger button is announced, which will introduce a regression losing the announcement of the trigger button's "toggling" nature.

**Comment by joshblack** [comment]
> introduce a regression losing the announcement of the trigger button's "toggling" nature.

This is why I said this above:

> Screen reader software would be able to say whether the listbox is open/closed through aria-haspopup and aria-expanded if I understand correctly.

As a result, the idea was that this would not need to be communicated through `aria-label`.

**Comment by asudoh** [comment]
Seems that you have a specific change you have in your mind, or specific lines in this PR that you think should be removed. Do you want to put the code snippet if it's the former? Do you want to put some links to the specific lines of code if it's the latter? Thanks!

**Comment by joshblack** [comment]
If I understand correctly, the diff would be:

- Remove `aria-label` from `bx--list-box__field` (already done)
- Associate `bx--list-box__field` with `bx--label` (already done) and `bx--list-box__label` through `aria-labelledby` so that the value is communicated
- Associate `bx--form__helper-text` with `bx--list-box__field` through `aria-describedby`

Could easily be wrong though, I believe the structure we have in general is inverted (e.g. trigger is a descendant of listbox, two nodes with `role="listbox"`, etc.) and we need to correct those as well for ARIA to work as intended 👍 This also changes between different types (listbox vs combobox, etc)

Reference:

- https://pattern-library.dequelabs.com/components/selects
- https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html
- https://dequelabs.github.io/combobo/demo/

cc @dakahn just to confirm, don't want to steer anyone in the wrong direction here 👍

**Comment by asudoh** [comment]
> Associate `bx--list-box__field` with ... and `bx--list-box__label` through `aria-labelledby` so that the value is communicated

Did you mean a change to `<div class="bx--list-box__field">` adding `aria-labelledby` pointing to `<span class="bx--list-box__label">` or something else?

**Comment by joshblack** [comment]
Yup! `<div class="bx--list-box__field" aria-labelledby="bx-label-id list-box__label-id">`

**Comment by asudoh** [comment]
Above suggested changes are fine with me, updated.

**Comment by abbeyhrt** [comment]
Hey @asudoh! For the MultiSelect component, it seems like this PR gets rid of the "Accessible name does not match visible text" DAP violation but then creates a "An interactive/widget must have an accessible name" violation. Not sure if we want/need to address this is in this PR but figured i'd point it out!

**Comment by asudoh** [comment]
@abbeyhrt Thank you for testing - Not sure if it's a false positive or not, but fixed anyway.

**Review by joshblack — APPROVED** [review]

**Comment by joshblack** [comment]
@abbeyhrt do you have a chance to re-review?? 👀

File: packages/react/src/components/Dropdown/Dropdown.js [hunk 1]
```diff
@@ -190,16 +190,24 @@ export default class Dropdown extends React.Component {
...
 
+    const helperId = !helperText
+      ? undefined
+      : `dropdown-helper-text-${this.dropdownInstanceId}`;
+    const labelId = `dropdown-label-${this.dropdownInstanceId}`;
+    const fieldLabelId = `dropdown-field-label-${this.dropdownInstanceId}`;
+
     const title = titleText ? (
-      <label htmlFor={id} className={titleClasses}>
+      <span id={id} className={titleClasses}>
```
Comment (abbeyhrt) [inline packages/react/src/components/Dropdown/Dropdown.js]: It seems like the `id` being used here and in `ListBox` is causing a DAP violation: 
Screen Shot 2020-01-07 at 4 41 22 PM

Comment (asudoh) [reply packages/react/src/components/Dropdown/Dropdown.js]: Great catch @abbeyhrt! Was an error in rebasing. Fixed.

**Review by abbeyhrt — APPROVED** [review]
Looks great! 🎉

**Pull request merged**